internal/database/sync.go

```diff
@@ -368,3 +370,134 @@ func (r SyncRepo) DeleteSyncLockFile(ctx context.Context, apiKey string) bool {
+	}
+
+	return &etag, nil
+}
+
+// Get sync data and etag
+func (r SyncRepo) GetSyncDataAndETag(ctx context.Context, apiKey string) ([]byte, *string, error) {
+	var etag string
+	var data []byte
+
+	err := r.db.squirrel.
+		Select("data", "data_etag").
+		From("sync_data").
+		Where(sq.Eq{"user_api_key": apiKey}).
+		Limit(1).
+		RunWith(r.db.handler).
+		Scan(&data, &etag)
+
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, nil, nil
+		}
+		return nil, nil, errors.Wrap(err, "error executing query")
+	}
+
+	return data, &etag, nil
+}
+
+// Create or replace sync data, returns the new etag.
+func (r SyncRepo) SetSyncData(ctx context.Context, apiKey string, data []byte) (*string, error) {
+	now := time.Now()
+	// the better way is use hash like sha1
+	// but uuid is faster than sha1
+	newEtag := "uuid=" + uuid.NewString()
+
+	updateResult, err := r.db.squirrel.
+		Update("sync_data").
+		Set("updated_at", now).
+		Set("data", data).
+		Set("data_etag", newEtag).
+		Where(sq.Eq{"user_api_key": apiKey}).
+		RunWith(r.db.handler).ExecContext(ctx)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "error executing query")
+	}
+
+	if rowsAffected, err := updateResult.RowsAffected(); err != nil {
+		return nil, errors.Wrap(err, "error executing query")
+	} else if rowsAffected == 0 {
+		// new item
+		insertResult, err := r.db.squirrel.
+			Insert("sync_data").
+			Columns(
+				"user_api_key",
+				"updated_at",
+				"data",
+				"data_etag",
+			).
+			Values(apiKey, now, data, newEtag).
+			RunWith(r.db.handler).ExecContext(ctx)
+
+		if err != nil {
+			return nil, errors.Wrap(err, "error executing query")
+		}
+
+		if rowsAffected, err := insertResult.RowsAffected(); err != nil {
+			return nil, errors.Wrap(err, "error executing query")
+
+		} else if rowsAffected == 0 {
+			// multi devices race condition
+			return nil, errors.New("no rows affected")
+		}
+	}
+
+	r.log.Debug().Msgf("Sync data upsert: api_key=\"%v\"", apiKey)
+	return &newEtag, nil
+}
+
+// Replace sync data only if the etag matches,
+// returns the new etag if updated, or nil if not.
+func (r SyncRepo) SetSyncDataIfMatch(ctx context.Context, apiKey string, etag string, data []byte) (*string, error) {
+	now := time.Now()
+	// the better way is use hash like sha1
+	// but uuid is faster than sha1
+	newEtag := "uuid=" + uuid.NewString()
+
+	result, err := r.db.squirrel.
+		Update("sync_data").
+		Set("updated_at", now).
+		Set("data", data).
+		Set("data_etag", newEtag).
+		Where(sq.Eq{"user_api_key": apiKey}).
+		Where(sq.Eq{"data_etag": etag}).
+		RunWith(r.db.handler).ExecContext(ctx)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "error executing query")
+	}
+
+	if rowsAffected, err := result.RowsAffected(); err != nil {
+		return nil, errors.Wrap(err, "error executing query")
+
+	} else if rowsAffected == 0 {
+		r.log.Debug().Msgf("Sync data replace fail due to etag not match: api_key=\"%v\", etag=\"%v\"", apiKey, etag)
```

Comment:
```suggestion
		r.log.Debug().Msgf("ETag mismatch for api_key=\"%v\". Concurrent modification detected, aborting sync to prevent data overwrite. ETag=\"%v\"", apiKey, etag)
```

Reply:
Maybe a better log message something like that or at least it's not a failure in this case.